async/mailbox.go

```diff
@@ -0,0 +1,119 @@
+package async
+
+// An AsyncMailbox stores AsyncErrors and their assocaited callbacks
+// and invokes them once the AsyncError is completed
+//
+// Often times we may spawn go routines in an event loop to do some concurrent work,
+// go routines provide no way to return a response, however we may want
+// to be notified if the work the go routine was doing completed succesfully
+// or unsuccessfully, and then take some action based on that result.
+// AsyncMailbox provides a construct to do this.
+//
+// The below example is a storeValue function, which tries to store
+// A value durably.  A value is considered durably stored if it successfully
+// writes to two of three replicas. We want to write to all replicas in parallel
+// and return as soon as two writes succeed. We return an error if < 2 writes succeed
+//
+//  func storeValue(num int) error {
+//    successfulWrites := 0
+//    returnedWrites := 0
+//    mailbox := NewAsyncMailbox()
+//
+//    writeCallback := func (err error) {
+//      if err != nil {
+//        successfulWrites++
+//      }
+//      returnedWrites++
+//    }
+//
+//    // Send to Replica One
```

Comment:
This means the async goroutine looks less like a function (instead of returning an error, it has to SetValue).

WWYT of mailbox.RunAsync(func() error) that does the wrapping? So it takes a function that looks normal, and then adds the async composition?